### PR TITLE
feat: Add first-class Comet ML experiment tracking

### LINKED ISSUE #2652: feat: Add first-class Comet ML experiment tracking
## Description

Megatron Bridge has first-class integrations for TensorBoard, Weights & Biases, and MLflow as experiment tracking backends. All three follow a consistent pattern: config fields in `LoggerConfig`, lazy-initialized properties in `GlobalState`, and symmetric inline logging calls in `training_log()`, `eval.py`, and `checkpointing.py`.

This issue proposes adding **Comet ML** as a fourth first-class backend, following the exact same structural pattern.

## Motivation

Users who use Comet ML for experiment tracking currently rely on Comet's TensorBoard auto-patcher (`COMET_AUTO_LOG_TENSORBOARD=1`), which:

- Only captures metrics written to TensorBoard, missing WandB-only or MLflow-only metrics
- Depends on Comet intercepting `SummaryWriter` calls, which is fragile
- Does not capture config/hyperparameters as Comet parameters
- Does not support checkpoint artifact tracking
- Does not integrate with the DLFw tensor inspect system

A first-class integration would provide the same level of metrics, parameters, and artifact logging that WandB and MLflow users already have.

## Proposed Changes

Following the existing WandB/MLflow pattern:

1. **Config fields** — Add `comet_project`, `comet_experiment_name`, `comet_workspace`, `comet_api_key`, `comet_tags` to `LoggerConfig`
2. **Lazy init** — Add `GlobalState.comet_logger` property (last-rank init, stores `comet_ml.Experiment` instance)
3. **Training metrics** — Add Comet logging calls alongside existing WandB/MLflow calls in `training_log()`
4. **Validation metrics** — Add Comet logging in `eval.py`
5. **Timer metrics** — Add `_timers_write_to_comet` monkey-patch
6. **Checkpoint tracking** — Add `comet_utils.py` with save/load callbacks
7. **Tensor inspect** — Add `_CometExperimentLogger` wrapper
8. **Run plugin** — Add `CometPlugin` for NeMo Run launching
9. **Cleanup** — Add `experiment.end()` at shutdown
10. **Documentation** — Add Comet section to `docs/training/logging.md`

## Config Example

```python
cfg.logger.comet_project = "my-project"
cfg.logger.comet_experiment_name = "qwen3-14b-sft"
cfg.logger.comet_workspace = "my-workspace"
cfg.logger.comet_tags = ["sft", "qwen3"]
```

## Reference

[NeMo Automodel PR #1411](https://github.com/NVIDIA-NeMo/Automodel/pull/1411) adds a similar Comet ML integration to the Automodel training recipe.

### PR DESCRIPTION
**Pull request by shanecmoran**
## Summary

Adds Comet ML as a fourth experiment tracking backend alongside TensorBoard, Weights & Biases, and MLflow. The integration follows the exact same structural pattern used by the existing backends — no new abstractions, just symmetric extension of each logging call site.

**Reference:** [NeMo Automodel PR #1411](https://github.com/NVIDIA-NeMo/Automodel/pull/1411) adds a similar Comet integration to Automodel.

### Config (`LoggerConfig`)
- 5 new fields: `comet_project`, `comet_experiment_name`, `comet_workspace`, `comet_api_key`, `comet_tags`
- `finalize()` validation: requires `comet_experiment_name` when `comet_project` is set, checks `comet_ml` importable

### Core (`GlobalState`)
- `comet_logger` lazy property: initializes `comet_ml.Experiment` on last rank, logs full config as parameters
- `_timers_write_to_comet` timer patch (preserves `/` in metric names unlike MLflow)
- Reset in `reset_for_restart()`

### Metric Logging
- `training_log()`: ~18 inline `if comet_logger:` blocks mirroring every `if mlflow_logger:` block
- `eval.py`: validation loss and PPL
- `train.py`: `experiment.end()` at both shutdown paths

### Checkpoint Tracking
- New `comet_utils.py` with `on_save_checkpoint_success` / `on_load_checkpoint_success`
- Wired into `checkpointing.py` save/load paths

### Integration Points
- `tensor_inspect.py`: `_CometExperimentLogger` wrapper for NVIDIA DLFw Inspect
- `run_plugins.py`: `CometPlugin` for NeMo Run launching
- `docs/training/logging.md`: full documentation section

### Tests
- `test_comet_utils.py`: save/load checkpoint callback tests
- `test_state.py`: comet_logger property, timer patch, reset tests
- `test_config.py`: finalize validation tests

## Config Example

```python
cfg.logger = LoggerConfig(
    tensorboard_dir="./runs/tensorboard",
    comet_project="my_project",
    comet_experiment_name="qwen3_14b_sft",
    comet_workspace="my_team",
    comet_tags=["sft", "qwen3"],
)
```

Fixes #2652

## Summary by CodeRabbit

* **New Features**
  * Added Comet ML logging integration for training workflows, enabling automatic logging of training metrics, validation performance, checkpoint metadata, and system telemetry.
  * New configuration options for Comet project, experiment name, workspace, API key, and custom tags.
  * Added comprehensive documentation for Comet ML setup and usage.

* **Tests**
  * Added unit tests validating Comet ML configuration, logging, and integration behavior.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -89,6 +89,7 @@ dependencies = [
     "timm",
     "open-clip-torch>=3.2.0",
     "mlflow>=3.5.0",
+    "comet-ml>=3.50.0",
```
Comment (yaoyu-33) [inline pyproject.toml:92]: I think uv.lock need update, otherwise ci will fail.

Comment (shanecmoran) [reply pyproject.toml:92]: Thanks — added comet-ml to pyproject.toml but can't regenerate uv.lock locally since it needs the CI environment for GPU-specific package resolution. Happy to update if there's a way to run the lock step externally.

Comment (yaoyu-33) [reply pyproject.toml:92]: shouldn't need ci env, just remove old lock file and re-run "uv lock" should work.
Or you can use our container / container Dockerfile mentioned in readme which will for sure work.

Comment (shanecmoran) [reply pyproject.toml:92]: Done — uv.lock has been updated and the branch has been rebased onto the latest main. Also added missing @pytest.mark.unit markers to the test classes in test_state.py. Ready for re-review when you get a chance.

**Comment by yaoyu-33** [comment]
we might need to take over this pr, convert this into a internal pr. uv lock update on 3rd party contribution is a bit hard to merge.

lemme know if it's okay.

**Comment by LoganVegnaSHOP** [comment]
@yaoyu-33 

Please feel free to take this on as an internal PR instead. Just keep me updated with the progress. I am taking on the Megatron Work from Shane at Shopify BTW
